### PR TITLE
[BEAM-7513] Bigquery rowcount

### PR DESCRIPTION
**Pull request by riazela**
[BEAM-7513] Adding rowcount estimation for bigquery tables. This change creates a class to represent beam rowcount estimation. It has to pass the pipeline options in SQL transform path to BeamCalciteTable. Therefore, currently the pipeline options are passed to BeamCalciteTable in two different ways in SQLTransform path we get it as PipelineOption object. In JDBC path we get it as a Map<String, String> object.
A Jira Issue is created to unify both paths so that they pass the PipelineOptions itself. https://issues.apache.org/jira/projects/BEAM/issues/BEAM-7590

R: @akedin 
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by riazela** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by riazela** [comment]
run java postcommit

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/BeamCalciteTable.java [hunk 1]
```diff
@@ -38,32 +41,70 @@
...
-    return new BeamCalciteTable(table, ImmutableMap.of());
+    return new BeamCalciteTable(table, ImmutableMap.of(), null);
   }
 
   @Override
   public RelDataType getRowType(RelDataTypeFactory typeFactory) {
     return CalciteUtils.toCalciteRowType(this.beamTable.getSchema(), typeFactory);
   }
 
+  private PipelineOptions getPipelineOptions() {
```
Comment (akedin) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/BeamCalciteTable.java]: Is this the only change from #8822 ? Can you make it a separate commit?

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/BeamCalciteTable.java [hunk 2]
```diff
@@ -38,32 +41,70 @@
...
 /** Adapter from {@link BeamSqlTable} to a calcite Table. */
 public class BeamCalciteTable extends AbstractQueryableTable
     implements ModifiableTable, TranslatableTable {
   private final BeamSqlTable beamTable;
-  private final Map<String, String> pipelineOptions;
+  private final Map<String, String> pipelineOptionsMap;
+  private PipelineOptions pipelineOptions;
 
-  BeamCalciteTable(BeamSqlTable beamTable, Map<String, String> pipelineOptions) {
+  //  BeamCalciteTable(BeamSqlTable beamTable, Map<String, String> pipelineOptionsMap) {
```
Comment (akedin) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/BeamCalciteTable.java]: remove this

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/BeamCalciteTable.java [hunk 3]
```diff
@@ -38,32 +41,70 @@
...
+import org.apache.calcite.schema.Statistics;
 import org.apache.calcite.schema.TranslatableTable;
 
 /** Adapter from {@link BeamSqlTable} to a calcite Table. */
 public class BeamCalciteTable extends AbstractQueryableTable
     implements ModifiableTable, TranslatableTable {
   private final BeamSqlTable beamTable;
-  private final Map<String, String> pipelineOptions;
+  private final Map<String, String> pipelineOptionsMap;
+  private PipelineOptions pipelineOptions;
```
Comment (akedin) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/BeamCalciteTable.java]: add a comment here with a link to the jira to refactor this piece

**Comment by akedin** [comment]
run direct runner nexmark tests

**Comment by akedin** [comment]
run dataflow runner nexmark tests

**Comment by akedin** [comment]
run flink runner nexmark tests

**Comment by akedin** [comment]
run spark runner nexmark tests

**Comment by akedin** [comment]
run sql postcommit

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/BeamCalciteTable.java (hunk 1 above)
Comment (riazela) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/BeamCalciteTable.java]: No. The pipelineoptions object was not passed to BeamCalciteTable. So the process in JDBC Driver that was converting pipeline options to map is also deleted. So JDBC Connection and JDBC Driver are also changed.

**Comment by akedin** [comment]
run sql postcommit

**Comment by akedin** [comment]
run java postcommit

**Comment by akedin** [comment]
run direct runner nexmark tests

**Comment by akedin** [comment]
run dataflow runner nexmark tests

**Comment by akedin** [comment]
run flink runner nexmark tests

**Comment by akedin** [comment]
run spark runner nexmark tests

**Comment by akedin** [comment]
LGTM. This time all seems good
